src/Symfony/Component/HttpClient/NativeHttpClient.php

```diff
@@ -110,10 +110,25 @@ public function request(string $method, string $url, array $options = []): Respo
             'debug' => \extension_loaded('curl') ? '' : "* Enable the curl extension for better performance\n",
         ];
 
+        $checkMaxDuration = null;
+        if (null !== $maxDuration = $options['max_duration']) {
+            $checkMaxDuration = static function () use ($maxDuration, &$info): void {
+                if ($info['total_time'] >= $maxDuration) {
+                    throw new TransportException(sprintf('Max duration was reached for "%s".', implode('', $info['url'])));
+                }
+            };
+        }
+
         if ($onProgress = $options['on_progress']) {
             // Memoize the last progress to ease calling the callback periodically when no network transfer happens
             $lastProgress = [0, 0];
-            $onProgress = static function (...$progress) use ($onProgress, &$lastProgress, &$info) {
+            $onProgress = static function (...$progress) use ($onProgress, &$lastProgress, &$info, $checkMaxDuration) {
+                if ($checkMaxDuration) {
+                    $info['total_time'] = microtime(true) - $info['start_time'];
```

Comment:
should be removed, total_time should already be up to date here
a similar line is missing in `CurlResponse::perform()` before calling this very callback btw

Reply:
👌 
> a similar line is missing in CurlResponse::perform() before calling this very callback btw

To fix on 4.3 I guess?

Reply:
on 4.3 yes please